### PR TITLE
[WIP] Move InteractionBox to Base Layer

### PR DESCRIPTION
**Pull request by jojoelfe**
# Description
This is very much a premature PR to clarify some things. Right now it only contains the model and some tests. The plan for how it should behave is like this:

The base layer creates an instance of this class:
```
self._interaction_box = InteractionBox()
```
A layer can then use it either for a dragging select behavior like this
```
self._interaction_box.points = [drag_start, drag_end]
self._interaction_box.show_handle = False
self._interaction_box.show = True
```
or use it for indicating selection and translate, scale, rotate behavior
```
self._interaction_box.points = [list of selected points]
self._interaction_box.show_handle = True
self._interaction_box.show = True
```

A few questions:
- My feeling is this class should handle mouse events and needs to listen to mouse_press, _move, etc.. . My initial feeling was to call corresponding methods from the mouse_* methods of the base layer, but these will be overwritten by the subclasses, right?
- If the box gets rotated or moved it probably should communicate this to the layer with events. The correct thing is to give it a .events EmitterGroup property and have layers connect to them?
- Rendering ideally should be done in vispy_base_layer.py? Then there needs to be a way of creating a Marker and Line visual in the base layer. We could do this with:
```
self.node = Compound([node, Line(), Markers()])
```
but this seems to create a lot of issues in existing code that expects self.node to be a specific visual.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

# References
This is based on discussion in #989 

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
Thanks @jojoelfe - sorry for not getting to this yet. It looks very exciting and that you're on the right track. I only have time to give some quick pointers right now, but will try and do a deeper dive on Friday.

> My feeling is this class should handle mouse events and needs to listen to mouse_press, move, etc.. . My initial feeling was to call corresponding methods from the mouse* methods of the base layer, but these will be overwritten by the subclasses, right?

Yeah, we have some plans to do some refactoring to actually make this easier in #694. I can also try and work on this over the weekend, which I think will help here.

> If the box gets rotated or moved it probably should communicate this to the layer with events. The correct thing is to give it a .events EmitterGroup property and have layers connect to them?

This sounds about good, and consistent with our current practices.

> Rendering ideally should be done in vispy_base_layer.py? Then there needs to be a way of creating a Marker and Line visual in the base layer. We could do this with:
`self.node = Compound([node, Line(), Markers()])`
but this seems to create a lot of issues in existing code that expects self.node to be a specific visual.

Yeah, I've been a little worried about that. The main problem I see is that even after the refactor some layer visuals have a single visual, but some will have two, one for the main visual and one for highlighting. Another way to think about this would be to have the `interaction box` visual get added only once to the canvas in the qt_viewer and then each layer can just update that visual with its interaction box, instead of the base layer having the visual and each layer having one. This will be much more performant too as if you have 10 layers instead of having ~30 visuals you'll still have ~10 visuals. Does that make sense? Do you think that sounds reasonable?

**Comment by jojoelfe** [comment]
@sofroniewn 

> Yeah, I've been a little worried about that. The main problem I see is that even after the refactor some layer visuals have a single visual, but some will have two, one for the main visual and one for highlighting. Another way to think about this would be to have the interaction box visual get added only once to the canvas in the qt_viewer and then each layer can just update that visual with its interaction box, instead of the base layer having the visual and each layer having one. This will be much more performant too as if you have 10 layers instead of having ~30 visuals you'll still have ~10 visuals. Does that make sense? Do you think that sounds reasonable?

That makes a lot of sense. I just pushed some code where the baselayer adds an additional node, which would results in the layer explosion you mentioned. It would be good to have a dedicated layer for this with an order of ```len(layers) + 1``` to keep it on top of everything else. 
The other thing I was worried about is that every layer would have to manually hide it's interaction box when it's not active. By having a single vispy visual that only renders the interaction box of the active layer we get that behaviour for free.

**Comment by sofroniewn** [comment]
> The other thing I was worried about is that every layer would have to manually hide it's interaction box when it's not active. By having a single vispy visual that only renders the interaction box of the active layer we get that behaviour for free.

Yes exactly - I'm imagining there will now be a separate `vispy_interaction_box.py` file and that will get instantiated once when the canvas is created around here in the code base:

https://github.com/napari/napari/blob/723c937d47c266aa37916c3ab5f06ca4daf7ffaf/napari/_qt/qt_viewer.py#L146-L150

Then when the active layer is changed the interaction box visual will get cleared / hidden / swapped to the interaction box of the new layer. That will happen in the `qt_viewer.py` too, but you might now need to add a new function that gets connected to the `active_layer` event that gets emitted from the viewer model here https://github.com/napari/napari/blob/723c937d47c266aa37916c3ab5f06ca4daf7ffaf/napari/components/viewer_model.py#L239

Note each I think that the actual interaction box model code can still live on the base model layer code as you have right now with the `_interaction_box`.

How does this sound?

I also want to give you a heads up that there is a proposal for refactoring our model / view separation in #1006 by adding an explicit controller. That will take a bit of time for the team to discuss and if we do decide to do it will take a bit of time to implement, so for now I'd recommend keeping on going with the current architecture, but it might be worth looking #1006 over too just to keep those ideas in mind.

**Comment by sofroniewn** [comment]
This is looking great @jojoelfe, I really like how the code is shaping up.

It looks like a big thing still missing at this point are the mouse functions. I think there we have some plans to generalize our mouse functions, see #793 at which point you could add mouse functions more easily to the interaction box (maybe using a mixin). For now though you might be best off adding them to the viewer and then having a quick way to determine if the interaction box is visible or not and whether they need to be called. We're moving towards the generator syntax described in #544 across the whole code base(see for this for the Labels layer for example in #1010, which is still in progress) and it would be nice to use that syntax if possible.

Thinking more about

> If the box gets rotated or moved it probably should communicate this to the layer with events. The correct thing is to give it a .events EmitterGroup property and have layers connect to them?

I think that making the connections directly inside the layer for now is fine, but that would be something that would go inside the controller if we move forward with #1006 I think (right @shanaxel42?) and could look much cleaner then. 

I'm curious how you're thinking about the scope of this PR and whether you'll actually start using the interaction box in the shapes layer in this PR or wait for a followup PR to hook it up there and remove the duplicate code? It might be easier for the rest of the team to review this PR if all it does is add the interaction box and we do the shapes stuff in a follow-up. If that is the case you might want to think about adding a small little example that forces the interaction box to display as it would be nice to have a way to see it work. You could add some simple callbacks in the example to print things out too.

**Comment by jojoelfe** [comment]
> We're moving towards the generator syntax described in #544 across the whole code base(see for this for the Labels layer for example in #1010, which is still in progress) and it would be nice to use that syntax if possible.

Yes, my next plan was to get the mouse events hooked up. I'll try to do it with the generator functions. 

> I think that making the connections directly inside the layer for now is fine, but that would be something that would go inside the controller if we move forward with #1006 I think (right @shanaxel42?) and could look much cleaner then.

I'll keep an eye out for #1006 and see how things develop. 

> I'm curious how you're thinking about the scope of this PR and whether you'll actually start using the interaction box in the shapes layer in this PR or wait for a followup PR to hook it up there and remove the duplicate code? It might be easier for the rest of the team to review this PR if all it does is add the interaction box and we do the shapes stuff in a follow-up. If that is the case you might want to think about adding a small little example that forces the interaction box to display as it would be nice to have a way to see it work. You could add some simple callbacks in the example to print things out too.

I think it makes sense to keep this PR to the API generation for the interaction_box and then have a separate PR that makes the shapes (and maybe points) layer use it. The example sounds like a good idea to show how things behave.

File: napari/layers/interaction_box.py [hunk 1]
```diff
@@ -1,9 +1,44 @@
 import numpy as np
+from skimage.transform import SimilarityTransform
 
 from ._constants import Box
 from ..utils.event import EmitterGroup, Event
 
 
+def inside_boxes(boxes):
```
Comment (jojoelfe) [inline napari/layers/interaction_box.py:8]: I copied this function from shape_utils.py, because I couldn't import it. As far as I can see this function is not used anywhere else.

Comment (sofroniewn) [reply napari/layers/interaction_box.py:8]: That's fine, we can remove from Shapes in follow up PR

**Comment by sofroniewn** [comment]
Hi @jojoelfe - apologies it's been so long getting back to you on this one. Maybe @andy-sweet can pick this up and do some review of it and we can decide how to move forward. As for context @andy-sweet some of the original motivation was to enable functionality like #989 which will now be much easier as we have a transforms class.

This interaction box could also be the foundation for a highlight layer - see discussion in #2560, #2954. #772 also has additional context that might help - cc @nclack

**Comment by andy-sweet** [comment]
Hey @jojoelfe - the idea and functionality look great here, so I'm happy to help out however I can.

I took a quick look tonight and also tried to get `examples/interaction_box.py` working.

At first, I attempted to create an environment based on the state of the repo at the last commit of this PR. But I got a vispy dependency error that didn't immediately make sense.

So I decided to merge main into this and handle that pain instead. There aren't many automatic conflicts, but there are plenty of updates needed to get that example working again. I made enough updates so that calling `napari examples/interaction_box.py`, but when I actually start interaction with viewer I get new ones.

I'm happy to keep working on resolve those until things seem fairly functional. But if you still have interest working on this, then it might be quicker for you to resolve some of those. Or I'm also happy to find some time together to combine our powers.

**Comment by jojoelfe** [comment]
Hi @sofroniewn  and @andy-sweet ,

yes, sorry I also got distracted. But I am still very interested in getting this working. It might take me a few days to get up to speed to remember what I have done and combine it with changes since then, but if you are not in a super rush it's probably easier for me to merge the current main and get things working again.

**Comment by sofroniewn** [comment]
> It might take me a few days to get up to speed to remember what I have done and combine it with changes since then, but if you are not in a super rush it's probably easier for me to merge the current main and get things working again.

Awesome, yes! Great if you can pick this back up again when convenient for you - no rush though. @andy-sweet can then help with the review and make sure we get it merged this time around.

Also as a note for #989 this should be much much easier now that we've added support for affine transforms on all our layers

**Comment by jojoelfe** [comment]
Hi @andy-sweet, I merged with main and caught up with enough changes to the coordinate system and events to get the example running again. I haven't really thought about whether the design still fits after these changes, but you can give it a try.

The general idea was for the interaction box two have two modes:

- A selection mode where due to a mouse-drag a box is drawn and events in the active layer are triggered, which can then decide whether items should be selected

- A transformation mode where the box is always drawn and has handles for rotation+scaling. When the box is then transformed it maintains an affine transformation which then gets passed to the active layer with events. The active layer can then decide whether to apply the transform to the complete layer (for images) or only selected items (for points)

The example only covers the first mode for now.

**Comment by andy-sweet** [comment]
That was quick @jojoelfe! I'll try to play with and review this on Monday, so expect feedback by the end of Monday or early Tuesday. Hope that's OK!

File: examples/interaction_box.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def on_selection_box_drag(event):
+    sel_i = points_in_box(event.box,viewer.layers.selection.active._view_data,viewer.layers.selection.active._view_size)
+    viewer.layers.selection.active.selected_data = sel_i
+    
+def on_selection_box_final(event):
+    sel_i = points_in_box(event.box,viewer.layers.selection.active._view_data,viewer.layers.selection.active._view_size)
+    viewer.layers.selection.active.selected_data = sel_i
+    viewer.layers.selection.active._interaction_box.points = np.array([viewer.layers.selection.active._view_data[i] for i in sel_i])
```
Comment (andy-sweet) [inline examples/interaction_box.py:18]:
```suggestion
    viewer.layers.selection.active._interaction_box.points = viewer.layers.selection.active._view_data[sel_i]
```

If the interaction box needs a copy of the data, you can wrap the RHS in `np.array(...)`.

There are some other similar changes in this file, so if this looks good, please make those too.

File: napari/_qt/qt_viewer.py [hunk 3]
```diff
@@ -232,6 +233,9 @@ def __init__(self, viewer: Viewer, show_welcome_screen: bool = False):
 
         # Add axes, scale bar
         self._add_visuals()
+        self._interaction_box_visual = VispyInteractionBox(viewer)
+        self._interaction_box_visual.node.parent = self.view.scene
+        self._interaction_box_visual.node.order = len(self.viewer.layers) + 1
```
Comment (andy-sweet) [inline napari/_qt/qt_viewer.py:238]: Is there a node with order `len(self.viewer.layers)`? If not, could we remove the `+ 1`?

File: napari/_vispy/_tests/test_vispy_interaction_box.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    qtbot.addWidget(view)
+
+    np.random.seed(0)
+    data = np.random.random((10, 15, 8))
+    layer = viewer.add_image(data)
+    layer._interaction_box.show = True
+    layer._interaction_box.points = np.array([[0, 0], [100, 100]])
+
+    assert (
+        view._interaction_box_visual.marker_node._data['a_position'].shape[0]
```
Comment (andy-sweet) [inline napari/_vispy/_tests/test_vispy_interaction_box.py:18]: Style/Nit: extract this as a local variable to make the assertions easier to read. E.g.

`marker_pos = view._interaction_box_visual.marker_node._data['a_position']`

File: napari/_vispy/_tests/test_vispy_interaction_box.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        == 9
+    )
+
+    viewer.add_image(data)
+
+    assert (
+        view._interaction_box_visual.marker_node._data['a_position'].shape[0]
```
Comment (andy-sweet) [inline napari/_vispy/_tests/test_vispy_interaction_box.py:25]: Can this be `marker_node.pos` instead?

File: napari/_vispy/_tests/test_vispy_interaction_box.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        == 1
```
Comment (andy-sweet) [inline napari/_vispy/_tests/test_vispy_interaction_box.py:26]: Why 1 instead of 9 here? I thought the 9 came from the 9 marker control points of the interaction box (rather than being specific to the points layer) - is that right? Did adding an image make the interaction box effectively hidden?

File: napari/_vispy/vispy_interaction_box.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            edge_color,
+            pos,
+            width,
+        ) = self.layer._interaction_box._compute_vertices_and_box()
+
+        if vertices is None or len(vertices) == 0:
+            vertices = np.zeros((1, self._viewer.dims.ndisplay))
+            size = 0
+        else:
+            vertices = vertices + 0.5
```
Comment (andy-sweet) [inline napari/_vispy/vispy_interaction_box.py:60]: Why is this 0.5 needed? Can you add an implementation comment to explain?

File: napari/layers/_constants.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+class Box:
+    """Box: Constants associated with the vertices of the interaction box
+    """
+
+    WITH_HANDLE = [0, 1, 2, 3, 4, 5, 6, 7, 9]
+    LINE_HANDLE = [7, 6, 4, 2, 0, 7, 9]
```
Comment (andy-sweet) [inline napari/layers/_constants.py:6]: I assume that 7 is in the list twice for a deliberate repeat? Consider adding an implementation comment explain why that repeat is needed.

File: napari/layers/_constants.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+class Box:
+    """Box: Constants associated with the vertices of the interaction box
```
Comment (andy-sweet) [inline napari/layers/_constants.py:2]: Would be good to clarify the integer values are indices of the vertices.

File: napari/layers/_constants.py (hunk 8 above)
Comment (andy-sweet) [reply napari/layers/_constants.py:6]: Also, does order matter here? This seems to go clockwise around the box, whereas the others go anti-clockwise.

File: napari/layers/_constants.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    WITHOUT_HANDLE = [0, 2, 4, 6]
+    LINE = [0, 2, 4, 6, 0]
+    TOP_LEFT = 0
+    TOP_CENTER = 7
+    LEFT_CENTER = 1
+    BOTTOM_RIGHT = 4
+    BOTTOM_LEFT = 2
+    CENTER = 8
+    HANDLE = 9
+    LEN = 8
```
Comment (andy-sweet) [inline napari/layers/_constants.py:16]: Is this supposed to be the number of vertices (i.e. is `LEN` like `len()` and short for length)? If so, shouldn't it be 10? If I misunderstood it might be good to rename this or add an explanatory implementation comment.

But, if it is meant to be the number of vertices, consider defining an Enum like `BoxVertex` (or maybe `BoxVertexIndex`) with just the single indices. Then you can access that as `len(BoxVertex)`. Then maybe you could define `BoxVertexGroup` for the others.

File: napari/layers/_constants.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+class Box:
+    """Box: Constants associated with the vertices of the interaction box
+    """
+
+    WITH_HANDLE = [0, 1, 2, 3, 4, 5, 6, 7, 9]
+    LINE_HANDLE = [7, 6, 4, 2, 0, 7, 9]
+    WITHOUT_HANDLE = [0, 2, 4, 6]
+    LINE = [0, 2, 4, 6, 0]
+    TOP_LEFT = 0
```
Comment (andy-sweet) [inline napari/layers/_constants.py:9]: It would be easier to understand the groups like `WITHOUT_HANDLE` if all single vertices were defined by name and integer index first. I.e. put the following first:

```python
TOP_LEFT = 0
LEFT_CENTER = 1
...
HANDLE = 9
```

then

```python
WITHOUT_HANDLE = [0, 2, 4, 6]
LINE = WITHOUT_HANDLE + [0]
```

You might also want to consider reusing the individual vertex names like
```python
WITHOUT_HANDLE = [TOP_LEFT, BOTTOM_LEFT, BOTTOM_RIGHT, TOP_RIGHT]
```

though that might be a bit verbose.

File: napari/layers/_tests/test_interaction_box.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
+import numpy as np
+
+from ..interaction_box import InteractionBox
+
+
+def test_interaction_box_initialization():
```
Comment (andy-sweet) [inline napari/layers/_tests/test_interaction_box.py:6]: I don't think we're consistent across the napari tests with stuff like this, but it would be good to rename this as the shorter `test_init` because it's already in a test module/suite/file that includes `interaction_box`.

Also, I prefer making each test capture one behavior, so that specific changes in behavior can be easily identified. So I would personally prefer to see this split up into a few tests with names like `test_init`, `test_init_without_handle`, `test_init_hide`, etc.

File: napari/layers/interaction_box.py [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+        self._show_vertices = show_vertices
+        self.events.points_changed()
+
+    @property
+    def angle(self):
+        offset = self._box[Box.HANDLE] - self._box[Box.CENTER]
+        angle = -np.degrees(np.arctan2(offset[0], -offset[1])) - 90
+        return angle
+
+    def points_in_box(self, points):
```
Comment (andy-sweet) [inline napari/layers/interaction_box.py:131]: This implementation is empty. I guess we plan to use the one in `_points_utils.py` instead?

File: napari/layers/interaction_box.py [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+            else:
+                box = self._box[Box.WITHOUT_HANDLE]
+
+            if self._selected_vertex is None:
+                face_color = 'white'
+            else:
+                face_color = self._highlight_color
+
+            edge_color = self._highlight_color
+            vertices = box[:, ::-1]
```
Comment (andy-sweet) [inline napari/layers/interaction_box.py:173]: Either remove this or the `if` statement. Or consider a ternary statement like

```python
vertices = box[:, ::-1] if self._show_vertices else np.empty((0, 2))
```

**Review by andy-sweet — COMMENTED** [review]
I'm still looking through most of the guts of `vispy_interaction_box.py`, so I'm only half way through of my review, but I thought I'd leave my comments for now, so you can act on those if you want.

In general, the functionality here is really nice and I liked playing around with the example. I did run into a couple of general issues while doing that, but have not identified the cause yet.
1. If you set the points layer to have a rotation (i.e. run `viewer.layers[0].rotate = 45`) after initializing the example, the interaction does not behave correctly. I'm not sure if that also occurs if you set the rotation when adding the layer.
2. At some point I got in a bad state where it seemed like nothing was selected, but trying to make a new selection by dragging caused some points to transform. I think I saw two boxes, that weren't always synced. But I'm not sure how to reproduce this - sorry!

I'm also still trying to understand how this will generalize to other layers. I think it would be good to add an image to the example, to understand how well that works.

**Comment by andy-sweet** [comment]
Also, I will try to finish this review tomorrow - sorry about the delay!

**Comment by jojoelfe** [comment]
Hi @andy-sweet,

thanks for the review so far. My thoughts about you points are:

1. I think this is be due to the fact that I have to make a decision whether the interaction box internally operates in canvas, world, or layer coordinates and then where the ultimate conversion into layer coordinates happens. My initial guess is that the InteractionBox should operate in canvas coordinates, since like the canvas it fundamentally operates in 2D and not in nD. Then there would have to be layer and viewer-mode specific code to convert into layer coordinates.

2. I encountered 2 scenarios where weird things happen. (a) When a single point is selected which produces a box with size 0 (b) When the layer mode is changed, which should enable or disable the InteractionBox, but isn't yet.

Also I just noticed that there are now overlays in the vispy code (for scalebar and axes). It seems to me like this would be a good place for the InteractionBox too and would get rid of the "len(layers) + 1" order you wondered about.

I will go ahead and make an example for the image layer.

Johannes

File: napari/layers/interaction_box.py [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+    c2 = ABAM <= ABAB
+    c3 = 0 <= BCBM
+    c4 = BCBM <= BCBC
+
+    inside = np.all(np.array([c1, c2, c3, c4]), axis=0)
+
+    return inside
+
+
+class InteractionBox:
```
Comment (sofroniewn) [inline napari/layers/interaction_box.py:42]: Could you make this a pydantic evented model like the ScaleBar & Axes. It means you wouldn't need to add the `EmitterGroup` yourself as that would come for free

**Review by sofroniewn — COMMENTED** [review]
> Also I just noticed that there are now overlays in the vispy code (for scalebar and axes). It seems to me like this would be a good place for the InteractionBox too and would get rid of the "len(layers) + 1" order you wondered about.

Yes, I like the idea of treating the interaction box like an overlay. My apologies for not mentioning this above - but I think that might be a better design. In that case I guess the box doesn't live on the layer itself, but lives on the viewer. It might be good to make a "highlight" object on the viewer and then have the interaction box inside there, so `viewer.highlights.interaction_box` or `viewer.overlays.interaction_box` and move scale bar/ axes over there too. We could add `viewer.interaction_box` to start too (@tlambert03 @andy-sweet any preference here?)

File: napari/layers/interaction_box.py (hunk 15 above)
Comment (sofroniewn) [reply napari/layers/interaction_box.py:42]: https://github.com/napari/napari/blob/3776819578ecf2785aaed4c0f464658609ea5272/napari/components/axes.py#L4

**Comment by sofroniewn** [comment]
I'll add I like `viewer.highlights.interaction_box` or `viewer.overlays.interaction_box` as we could put a ruler here too cc @nclack, it just seems better than putting them all on the viewer directly

File: napari/layers/interaction_box.py [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+        box = self._box
+
+        if box is not None:
+            rot = box[Box.TOP_CENTER]
+            length_box = np.linalg.norm(
+                box[Box.BOTTOM_LEFT] - box[Box.TOP_LEFT]
+            )
+            if length_box > 0:
+                r = self._rotation_handle_length
+                rot = (
```
Comment (andy-sweet) [inline napari/layers/interaction_box.py:208]: The formatting here makes the operation a little hard to parse.

Maybe the best thing to do here is to store a local variable and do something like

```python
left_edge = box[Box.BOTTOM_LEFT] - box[Box.TOP_LEFT]
height = np.linalg.norm(left_edge)
...
rot -= r * left_edge / height
```

File: napari/layers/interaction_box.py [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+        self.show = False
+        self.show_handle = True
+        self.show_vertices = True
+        # self.points = None
+        if self._box is not None:
+            self.events.selection_box_changed_final(
+                box=self._box[Box.WITHOUT_HANDLE]
+            )
+
+    def initialize_mouse_events(self, layer):
```
Comment (andy-sweet) [inline napari/layers/interaction_box.py:373]: The only use of this is immediately after creating `InteractionBox` in the base layer.

Instead, could we make this private and just call this at the end of `__init__`. Or do we expect to use this elsewhere?

**Review by andy-sweet — COMMENTED** [review]
Sorry, I'm still making my way through `interaction_box.py`, but I had a couple more comments that might be useful, so I added those.

> My initial guess is that the InteractionBox should operate in canvas coordinates, since like the canvas it fundamentally operates in 2D and not in nD.

Yes, this sounds reasonable to me. Given that we often use `Layer.world_to_data` (which includes any grid offset transforms) to convert from `event.position` in a mouse related callback, I'm assuming that `event.position` are in the world coordinate system, but restricted to the displayed dimensions. Is that what you mean by canvas coordinates, or is there another transform I'm missing?

If those are canvas coordinates, then it seems like the box coordinates should always defined in that space, so translation, scale, rotation of the box is straightforward. When we want to transform the actual points data, I think we can apply the same transform to the points in canvas coordinates, then apply the `world_to_data` transform to define the new data (in data coordinates). I guess the other situation we care about is defining which points are within the box as it's initially defined - I think that's most easily done in the canvas coordinates because the selection box will be axis-aligned in that coordinate space.

**Comment by andy-sweet** [comment]
> It might be good to make a "highlight" object on the viewer and then have the interaction box inside there, so `viewer.highlights.interaction_box` or `viewer.overlays.interaction_box` and move scale bar/ axes over there too. We could add `viewer.interaction_box` to start too (@tlambert03 @andy-sweet any preference here?)

I don't have a strong preference, but I'm also not sure exactly what highlight means in this context. I think that highlights are used to describe selected visuals (e.g. a selected point has a visible `MarkerVisual` highlight). But maybe they are more generally used to describe things that also define those selections and are interactive? Whereas overlays are non-interactive?

**Comment by sofroniewn** [comment]
> I don't have a strong preference, but I'm also not sure exactly what highlight means in this context. I think that highlights are used to describe selected visuals (e.g. a selected point has a visible MarkerVisual highlight). But maybe they are more generally used to describe things that also define those selections and are interactive? Whereas overlays are non-interactive?

One thought might be if the highlight/ overlay doesn't depend on the layer data (i.e. is a bounding box or a ruler etc) then it should live on the viewer, but if it depends on the layer data (e.g. outlines of the selected shapes) then it should live on the layer. i don't know though, that's just a thought

**Comment by jojoelfe** [comment]
I have been refactoring this as an overlay and using an EventedModel from scratch in a separate branch. Would it be better to close this and open a new PR or should I force-push the commits into this branch?

**Comment by andy-sweet** [comment]
> I have been refactoring this as an overlay and using an EventedModel from scratch in a separate branch. Would it be better to close this and open a new PR or should I force-push the commits into this branch?

If it's a fairly different approach (and it sounds like it is), then I would probably open a new PR and close this one. You can always refer back to this PR if you resolved some things mentioned here.

But feel free to do either.

**Comment by jojoelfe** [comment]
I made a new pull-request for the refactor at #3577
